src/asus.c

```diff
@@ -0,0 +1,429 @@
+/*
+ * Copyright (C) 2021 Kyoken, kyoken@kyoken.ninja
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the next
+ * paragraph) shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include <math.h>
+
+#include "config.h"
+#include "libratbag-private.h"
+#include "libratbag-data.h"
+
+#include "asus-defines.h"
+#include "asus-common.h"
+
+#define ASUS_CMD_GET_LED_DATA		0x0312  // get all leds
+#define ASUS_CMD_GET_SETTINGS		0x0412  // dpi, rate, button response, angle snapping
+#define ASUS_CMD_GET_BUTTON_DATA	0x0512  // get all buttons
+#define ASUS_CMD_GET_PROFILE_DATA	0x0012  // get current profile info
+#define ASUS_CMD_SET_LED		0x2851  // set single led
+#define ASUS_CMD_SET_SETTING		0x3151  // dpi / rate / button response / angle snapping
+#define ASUS_CMD_SET_BUTTON		0x2151  // set single button
+#define ASUS_CMD_SET_PROFILE		0x0250  // switch profile
+#define ASUS_CMD_SAVE			0x0350  // save settings
+
+static unsigned int ASUS_LED_TYPE[] = {
+	RATBAG_LED_TYPE_LOGO,
+	RATBAG_LED_TYPE_WHEEL,
+	RATBAG_LED_TYPE_SIDE,  // bottom back
+};
+
+// search for asus button index by ratbag types
+int
+asus_find_button_index_by_action(struct ratbag_button_action action)
+{
+	for (unsigned int i = 0; i < ARRAY_LENGTH(ASUS_BUTTON_MAPPING); i++) {
```

Comment:
just a btw, we also have `ARRAY_FOR_EACH` which would work here. No action needed if you don't want to change it.